Correct the type of default and await the value

cheese-stands-alone opened the pull request:
<!-- THANK YOU for your contribution to Dynamoose, we really appreciate you taking the time to improve this package, and look forward to reviewing your PR and getting the changes integrated into the package. Thanks again!! -->


### Summary:
According to the [docs](https://github.com/dynamoose/dynamoose/blob/v4.0.3/packages/dynamoose/lib/Schema.ts#L460) for a schema default value
`You can also pass in async functions or a function that returns a promise to the default property and Dynamoose will take care of waiting for the promise to resolve before saving the object.`

This is not reflected in the types for `default`, additionally the value is not currently not awaited. This PR corrects the types and fixes the logic to match the documentation.



<!-- Please remove the `Code sample` section below if it doesn't apply to this PR -->
### Code sample:
#### Schema
```
{"id": {"type": String, "default": async () => "Hello World"}}
```
In the above code sample, with his PR the default value will now be "Hello World" (previously it would be undefined)




### Type (select 1):
- [x] Bug fix
- [ ] Feature implementation
- [ ] Documentation improvement
- [ ] Testing improvement
<!-- If you select the option below, please replace `----` below with the issue number of the GitHub issue raised, and the user who asked you to submit a broken test -->
- [ ] Test added to report bug (GitHub issue #---- @---)
- [ ] Something not listed here


### Is this a breaking change? (select 1):
- [ ] 🚨 YES 🚨
- [x] No
- [ ] I'm not sure


### Is this ready to be merged into Dynamoose? (select 1):
- [x] Yes
- [ ] No


### Are all the tests currently passing on this PR? (select 1):
- [x] Yes
- [ ] No


### Other:
- [x] I have read through and followed the Contributing Guidelines
- [x] I have searched through the GitHub pull requests to ensure this PR has not already been submitted
- [x] I have updated the Dynamoose documentation (if required) given the changes I made
- [x] I have added/updated the Dynamoose test cases (if required) given the changes I made
- [x] I agree that all changes made in this pull request may be distributed and are made available in accordance with the [Dynamoose license](https://github.com/dynamoose/dynamoose/blob/main/LICENSE)
- [x] All of my commits and commit messages are detailed, explain what changes were made, and are easy to follow and understand
- [x] I have filled out all fields above

cheese-stands-alone commented:
I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

fishcharlie reviewed (changes requested):
Thanks for the PR! However it doesn't really look like you followed the guidelines for this project.

What is the purpose of this PR? It looks like possibly it fixes a bug where if you have an async function as a default function for a property it doesn't work as expected? But none of that was described in the PR description, and no linked issue exists for this PR, so it's hard to know the intent behind this.

Also it looks like you checked the box for saying you added automated tests for this PR, but there are no automated tests added. What is to prevent a regression in the future of this fix?

fishcharlie commented on packages/dynamoose/test/Schema.js:
```diff
@@ -1331,6 +1331,56 @@ describe("Schema", () => {
 			it(test.name, async () => {
 				const schema = new dynamoose.Schema(test.schema);
 				const output = await schema.getAttributeSettingValue(...test.input);
+				schema.defaultCheck;
```
@cheese-stands-alone I agree with Copilot here. Any reason for this line?

cheese-stands-alone replied:
Correct, that was accidentally added.

fishcharlie reviewed (changes requested):
@cheese-stands-alone I'm sorry for the continuous back and forth. I probably should have tested this before my last comment. So long as there aren't more problems in your future commits I promise this is the last thing.

So I ran the entire test suite and reverted `packages/dynamoose/lib/Schema.ts` line 1397 (basically removed the `await`). And all of the tests pass. I expect at least 1 test to fail when I revert that line, and all tests to pass with that fix you did. Otherwise we aren't really testing that change for regressions in the future. Right now all the tests pass even if I revert that change.

cheese-stands-alone commented:
@fishcharlie I updated the PR with the requested test cases. Of the 2 I just added the one about null/undfined makes logical sense to me. But the other one about arrays doesn't, but the test case is checking what I think the intended logic is. You might want to review the logic of packages/dynamoose/lib/Schema.ts:1397-1406.
